src/librustdoc/html/render/mod.rs

```diff
@@ -2410,3 +2420,160 @@ fn collect_paths_for_type(first_ty: clean::Type, cache: &Cache) -> Vec<String> {
+        format!(
+            r#"<a href="{root}{url}" target="_blank">{name}</a>"#,
+            root = cx.root_path(),
+            url = call_data.url,
+            name = call_data.display_name
+        )
+    };
+
+    // Generate the HTML for a single example, being the title and code block
+    let write_example = |w: &mut Buffer, (path, call_data): (&PathBuf, &CallData)| {
+        // FIXME(wcrichto): is there a better way to handle an I/O error than a panic?
+        //  When would such an error arise?
+        let contents =
+            fs::read_to_string(&path).expect(&format!("Failed to read file: {}", path.display()));
+
+        // To reduce file sizes, we only want to embed the source code needed to understand the example, not
+        // the entire file. So we find the smallest byte range that covers all items enclosing examples.
+        assert!(call_data.locations.len() > 0);
+        let min_loc =
+            call_data.locations.iter().min_by_key(|loc| loc.enclosing_item.byte_span.0).unwrap();
+        let min_byte = min_loc.enclosing_item.byte_span.0;
+        let min_line = min_loc.enclosing_item.line_span.0;
+        let max_byte =
+            call_data.locations.iter().map(|loc| loc.enclosing_item.byte_span.1).max().unwrap();
+
+        // The output code is limited to that byte range.
+        let contents_subset = &contents[(min_byte as usize)..(max_byte as usize)];
+
+        // The call locations need to be updated to reflect that the size of the program has changed.
+        // Specifically, the ranges are all subtracted by `min_byte` since that's the new zero point.
+        let (byte_ranges, line_ranges): (Vec<_>, Vec<_>) = call_data
+            .locations
+            .iter()
+            .map(|loc| {
+                let (byte_lo, byte_hi) = loc.call_expr.byte_span;
+                let (line_lo, line_hi) = loc.call_expr.line_span;
+                ((byte_lo - min_byte, byte_hi - min_byte), (line_lo - min_line, line_hi - min_line))
+            })
+            .unzip();
+
+        let edition = cx.shared.edition();
+        write!(
+            w,
+            r#"<div class="scraped-example" data-code="{code}" data-locs="{locations}">
+                <div class="scraped-example-title">{title}</div>
+                 <div class="code-wrapper">"#,
+            title = example_url(call_data),
+            // The code and locations are encoded as data attributes, so they can be read
+            // later by the JS for interactions.
+            code = contents_subset.replace("\"", "&quot;"),
+            locations = serde_json::to_string(&line_ranges).unwrap(),
+        );
+        write!(w, r#"<span class="prev">&pr;</span> <span class="next">&sc;</span>"#);
+        write!(w, r#"<span class="expand">&varr;</span>"#);
+
+        // FIXME(wcrichto): where should file_span and root_path come from?
+        let file_span = rustc_span::DUMMY_SP;
+        let root_path = "".to_string();
```

Comment:
I'm not sure about `file_span`, but for `root_path`, you could try looking at what the call to `print_src` in `src/librustdoc/html/source.rs` does.

Reply:
Figured it out.